### PR TITLE
Add catalog.bom application white/blacklists.

### PR DESCRIPTION
**Pull request by geomacy**
Note this PR depends on https://github.com/apache/brooklyn-server/pull/100 and should be merged after it.

It is desirable to have a config time mechanism to control what
bundles are allowed to present applications in the catalog and
what aren't.

This change adds a white/blacklist mechanism to the catalog.bom
bundle scanning. Each list takes the form of a CSV whitelist of
regexes to match against bundle symbolic ids.  A bundle will be
allowed to add an application (template) to the catalog if its
bundle id matches one of the regexes on the whitelist, and also
does not match one of the regexes on the blacklist.

The configuration keys and default values for the lists are:

brooklyn.catalog.osgi.application.whitelist=.*
brooklyn.catalog.osgi.application.blacklist=

(i.e. by default accept all bundles).

these values are specified in the file org.apache.brooklyn.core.catalog.bomscanner.cfg

Note that two 'default'  example applications are supplied in the bom
for brooklyn-software-base.

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 1]
```diff
@@ -74,6 +80,17 @@ public void unbind(ServiceReference<ManagementContext> managementContext) throws
 
     public class CatalogPopulator extends BundleTracker<Iterable<? extends CatalogItem<?, ?>>> {
 
+        private final String BROOKLYN_PACKAGE_REGEX = "org.apache.brooklyn.*";
+
+        private final ConfigKey<List<String>> WHITELIST =
+            ConfigKeys.newConfigKey(new TypeToken<List<String>>(){}, "brooklyn.catalog.osgi.application.whitelist",
+                "CSV Whitelist of patterns to match bundle symbolic id if bundle is to be permitted to add " +
+                    "applications (templates) to the catalog", ImmutableList.of(BROOKLYN_PACKAGE_REGEX));
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: If I am evaluating Brooklyn and load my own bundle in the container I'd expect that its `catalog.bom` is picked up out of the box. I think Brooklyn should accept all bundles, letting downstream distributions set their own policy.

Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: That's a very good point. Will update.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 2]
```diff
@@ -74,6 +80,17 @@ public void unbind(ServiceReference<ManagementContext> managementContext) throws
 
     public class CatalogPopulator extends BundleTracker<Iterable<? extends CatalogItem<?, ?>>> {
 
+        private final String BROOKLYN_PACKAGE_REGEX = "org.apache.brooklyn.*";
+
+        private final ConfigKey<List<String>> WHITELIST =
+            ConfigKeys.newConfigKey(new TypeToken<List<String>>(){}, "brooklyn.catalog.osgi.application.whitelist",
+                "CSV Whitelist of patterns to match bundle symbolic id if bundle is to be permitted to add " +
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Mention that regexes are accepted.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 3]
```diff
@@ -74,6 +80,17 @@ public void unbind(ServiceReference<ManagementContext> managementContext) throws
...
+                    "applications (templates) to the catalog", ImmutableList.of(BROOKLYN_PACKAGE_REGEX));
+        private final ConfigKey<List<String>> BLACKLIST =
+            ConfigKeys.newConfigKey(new TypeToken<List<String>>(){}, "brooklyn.catalog.osgi.application.blacklist",
+                "CSV Blacklist of patterns to match bundle symbolic id to prevent selected whitelisted bundles " +
+                    "adding applications (templates) to the catalog", ImmutableList.<String>of());
+
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Since this is a configuration set at dist build time do you think it will be better if set in [`org.apache.brooklyn.osgilauncher.cfg`](https://github.com/apache/brooklyn-server/blob/master/karaf/init/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L33)? 

`brooklyn.properties` is reserved for users, so they will have to add this key themselves.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 4]
```diff
@@ -137,9 +153,19 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
             }
         }
 
+        private void removeFromCatalog(CatalogItem<?, ?> item) {
+            try {
+                getManagementContext().getCatalog().deleteCatalogItem(item.getSymbolicName(), item.getVersion());
+            } catch (Exception e) {
+                LOG.warn("Failed to remove {} {} from catalog as {}", new String[] {
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: `propagateIfFatal` got lost in the rebase. Better insert it as the first expression after the catch.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 5]
```diff
@@ -121,12 +138,11 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
...
             }
             LOG.debug("Unloading catalog BOM entries from {} {} {}", bundleIds(bundle));
-            final BrooklynCatalog catalog = getManagementContext().getCatalog();
             for (CatalogItem<?, ?> item : items) {
                 LOG.debug("Unloading {} {} from catalog", item.getSymbolicName(), item.getVersion());
 
                 try {
-                    catalog.deleteCatalogItem(item.getSymbolicName(), item.getVersion());
+                    removeFromCatalog(item);
                 } catch (Exception e) {
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: try-catch not needed any more.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 6]
```diff
@@ -137,9 +153,19 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
             }
         }
 
+        private void removeFromCatalog(CatalogItem<?, ?> item) {
+            try {
+                getManagementContext().getCatalog().deleteCatalogItem(item.getSymbolicName(), item.getVersion());
+            } catch (Exception e) {
+                LOG.warn("Failed to remove {} {} from catalog as {}", new String[] {
+                    item.getSymbolicName(), item.getVersion(), e.getMessage()
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Why not pass the exception as a throwable argument to log, instead of including the message only. Could mask an underlying problem (for example NPE will have an empty message).

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 7]
```diff
@@ -155,9 +181,42 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
...
+            final List<String> white = managementContext.getConfig().getConfig(WHITELIST);
+            final List<String> black = managementContext.getConfig().getConfig(BLACKLIST);
+
+            return on(bundle, white) && !on(bundle, black);
+        }
+
+        private boolean on(Bundle bundle, List<String> list) {
+            for (String candidate : list) {
+                final String symbolicName = bundle.getSymbolicName();
+                if (symbolicName.matches(candidate)) {
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Do we need to trim the individual values? Or does that already happen in the coercion step? What if there's just only one item?

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 8]
```diff
@@ -155,9 +181,42 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
...
         }
 
+        private void removeAnyApplications(Iterable<? extends CatalogItem<?, ?>> catalogItems) {
+
+            final Iterator<? extends CatalogItem<?, ?>> cit = catalogItems.iterator();
+            while (cit.hasNext()) {
+                final CatalogItem<?, ?> item = cit.next();
+                if (TEMPLATE.equals(item.getCatalogItemType())) {
+                    removeFromCatalog(item);
+                    cit.remove();
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: (minor) No need for an iterator & removing the items from `catalogItems`. `removeFromCatalog` does all that's needed.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 9]
```diff
@@ -168,11 +227,18 @@ private String addLibraryDetails(Bundle bundle, String bomText) {
                     LOG.warn("Unexpected syntax for {} (expected Map), ignoring", BROOKLYN_CATALOG);
                 }
             }
-            final String updatedBom = new Yaml().dump(bom);
+            final String updatedBom = backToYaml(bom);
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Something that I didn't consider before when discussing this approach is that parsing and re-serializing the yaml will change its formatting. Not a problem for Brooklyn internals, but will be surprising to users when viewing the plan in the catalog. Not to be addressed in this PR, but we should keep it in mind.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java (hunk 3 above)
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: I guess it would be handy enough to add them in that file and then it could be referred to in the documentation (will do documentation as a separate PR).

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java (hunk 8 above)
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: This method modifies its argument, which is ultimately used as a return value for addingBundle, which becomes the set of items that is tracked by the BundleTracker, so if I remove something from the catalog in here, I also need to remove it from the iterable.  This probably this isn't clear enough from the code as written, so I shall change this to return a new Iterable rather than modify the arg.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java (hunk 7 above)
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: I'll trim the value, it's good to be explicit.  It should work fine with just one item.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java (hunk 9 above)
Comment (geomacy) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: That is true, but the formatting added here will hopefully be good enough looking to be acceptable.  Certainly worth keeping in mind.

**Comment by geomacy** [comment]
Code review comments done and commits squashed.

**Comment by geomacy** [comment]
added license header to cfg file and re-squashed

File: utils/common/src/main/java/org/apache/brooklyn/util/text/Strings.java [hunk 10]
```diff
@@ -916,4 +916,17 @@ public static String toUniqueString(Object x, int optionalMax) {
...
+
+    /**
+     * Parses a string as comma separated values and returns a list of the entries with whitespace removed.
+     * @param csv
+     * @param separators
+     */
+    public static List<String> parseCsv(String csv, String separators) {
+        List<String> result = MutableList.of();
+        for (String value: csv.split(separators)) {
+            result.add(value.trim());
```
Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/text/Strings.java]: Suggest adding some tests for this method. `.split` will return non-empty array for the empty string I believe

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java [hunk 11]
```diff
@@ -121,25 +157,27 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
...
+            }
+        }
+
+        private void removeFromCatalog(CatalogItem<?, ?> item) {
+            try {
+                getManagementContext().getCatalog().deleteCatalogItem(item.getSymbolicName(), item.getVersion());
+            } catch (Exception e) {
+                Exceptions.propagateIfFatal(e);
+                LOG.warn(Strings.join(new String[] {
+                    "Failed to remove", item.getSymbolicName(), item.getVersion(), "{} {} from catalog"
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java]: Remove `{} {}`

**Comment by neykov** [comment]
Looks great, a couple of minor points worth addressing.
